language-design.md

```diff
@@ -17,7 +17,11 @@ For each quote, describe what it was about the quote that led you pick it.
 
 **Response**
 
+In “Growing a Language” by Guy Steele, the author argues: “In a race, a small language with warts will beat a well designed language because users will not wait for the right thing; they will use the language that is quick and cheap, and put up with the warts. Once a small language fills a niche it is hard to take its place.” [Steele, 1998]. We found this interesting and compelling. When we dream about creating our own programming language, we want everything to be perfect. We are tempted to spend countless hours pouring over our code trying to make it perfect, scoffing at “bad programmers” who have create ugly, error-prone and unintuitive languages or programs. But when we consider this quote, it’s difficult to deny its truth. Creating a perfect programming language is impossible, and even making a very good language is in some ways doomed to failure. No one will wait for you to build the perfect system. Even if you succeed, by the time you do, people will have adopted a different language, become accustomed to its quirks, and found ways to work around them. They will not want to invest the time and energy in learning to use a “better” programming language if the language thy already have works well enough for their purposes.
 
+However, while designing the perfect programming language may be an impossible task, there are some good design principles that any successful programming language should follow. Joshua Bloch, writing specifically about API design writes: “**Obey the principle of least astonishment.** Every method should do the least surprising thing it could, given its name. If a method doesn’t do what users think it will, bugs will result.” [Bloch, 2006]. Most programmers’ immediate reaction to this quote is a small laugh and a murmured “Ugh, yes”. A function called max() must return the maximum number out of some list of numbers, if it doesn’t, users will assume it does, use it that way, and then spend hours searching their code for the bug. If they find it, they will only be more frustrated that a fault in the language made them waste so much time. Programming languages may seem cryptic to novices, but they follow general rules and guidelines that should be true across all languages even if they vary in the exact way they implement things. Programmers will not put up with languages that break these rules in too many ways.
```

Comment:
I definitely have a similar reaction to that of most programmers here. I hadn't even been thinking of this quote when I added my example of bad design in languages, but the fact that 0000 and 00 are different years in PHP (which is why one is an error and one is not) definitely violates this principle. Since I can't think of a good reason why PHP _should_ do that, I think your response here supports my example.